solr/core/src/java/org/apache/solr/search/MemAllowedLimit.java

```diff
@@ -0,0 +1,130 @@
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.solr.common.params.CommonParams;
+import org.apache.solr.request.SolrQueryRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MemAllowedLimit implements QueryLimit {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final double MEBI = 1024.0 * 1024.0;
+  private static final ThreadMXBean threadBean = ManagementFactory.getThreadMXBean();
+  private static Method GET_BYTES_METHOD;
+  private static boolean supported;
+
+  static {
+    boolean testSupported;
+    try {
+      Class<?> sunThreadBeanClz = Class.forName("com.sun.management.ThreadMXBean");
+      if (sunThreadBeanClz.isAssignableFrom(threadBean.getClass())) {
+        Method m = sunThreadBeanClz.getMethod("isThreadAllocatedMemorySupported");
+        Boolean supported = (Boolean) m.invoke(threadBean);
+        if (supported) {
+          m = sunThreadBeanClz.getMethod("setThreadAllocatedMemoryEnabled", boolean.class);
+          m.invoke(threadBean, Boolean.TRUE);
+          testSupported = true;
+          GET_BYTES_METHOD = sunThreadBeanClz.getMethod("getCurrentThreadAllocatedBytes");
+        } else {
+          testSupported = false;
+        }
+      } else {
+        testSupported = false;
+      }
+    } catch (Exception e) {
+      testSupported = false;
+    }
+    supported = testSupported;
+  }
+
+  private static final ThreadLocal<AtomicLong> threadLocalMem =
+      ThreadLocal.withInitial(() -> new AtomicLong(-1L));
+
+  private long limitBytes;
+  private AtomicLong accumulatedMem = new AtomicLong();
+  private long exitedAt = 0;
+
+  public MemAllowedLimit(SolrQueryRequest req) {
+    if (!supported) {
+      throw new IllegalArgumentException(
+          "Per-thread memory allocation monitoring not available in this JVM.");
+    }
+    float reqMemLimit = req.getParams().getFloat(CommonParams.MEM_ALLOWED, -1.0f);
+    if (reqMemLimit <= 0.0f) {
+      throw new IllegalArgumentException(
+          "Check for limit with hasMemLimit(req) before creating a MemAllowedLimit!");
+    }
+    limitBytes = Math.round(reqMemLimit * MEBI);
+  }
+
+  @VisibleForTesting
+  MemAllowedLimit(float memLimit) {
+    limitBytes = Math.round(memLimit * MEBI);
+  }
+
+  @VisibleForTesting
+  static boolean isSupported() {
+    return supported;
+  }
+
+  static boolean hasMemLimit(SolrQueryRequest req) {
+    return req.getParams().getFloat(CommonParams.MEM_ALLOWED, -1.0f) > 0.0f;
+  }
+
+  @Override
+  public boolean shouldExit() {
+    if (exitedAt > 0L) {
+      return true;
+    }
+
+    try {
+      long currentAllocatedBytes = (Long) GET_BYTES_METHOD.invoke(threadBean);
+      AtomicLong threadMem = threadLocalMem.get();
+      threadMem.compareAndExchange(-1L, currentAllocatedBytes);
+      long lastAllocatedBytes = threadMem.get();
+      accumulatedMem.addAndGet(currentAllocatedBytes - lastAllocatedBytes);
```

Comment:
On the first pass this appears to be zero, meaning that you are ignoring bytes allocated before the first invocation of the limit check? Is that the intent?

This class needs javadoc clearly stating what it intends to measure, so that future maintainers know what not to break (or if they want to break what you did on purpose). As it is I find it hard to evaluate the logic, and I am left shrugging and saying "I guess he meant to do that"

Reply:
> On the first pass this appears to be zero, meaning that you are ignoring bytes allocated before the first invocation of the limit check? Is that the intent?

Fixed. Added javadoc.